### DISCUSSION TITLE
Communication between opendds and fastdds Opendds sent data Fastdds did not receive it

### DISCUSSION THREAD

**Discussion by hong12234** [eProsima/Fast-DDS#3950] [Support]
### Is there an already existing issue for this?

- [X] I have searched the existing issues

### Expected behavior

192.168.225.32 send and 192.168.225.1 response，Communication between opendds and fastdds Opendds sent data Fastdds did not receive it，There were three interactions in this message, and the last interaction Fastdds did not receive any data。The figure shows the last packet of 192.168.225.32 data, but FASTDds did not receive the data。
![image](https://github.com/eProsima/Fast-DDS/assets/38304396/0e58ce80-5fa8-41f9-9547-d223cf874e8a)
The following figure shows：
![image](https://github.com/eProsima/Fast-DDS/assets/38304396/c3f8c889-59c4-41b1-a5dd-a9f2f6dca538)
Here are the captured network messages：

### Current behavior

/

### Steps to reproduce

/

### Fast DDS version/commit

fastdds 2.1

### Platform/Architecture

Other. Please specify in Additional context section.

### Transport layer

UDPv4

**Comment by hong12234** [comment]
[1.zip](https://github.com/eProsima/Fast-DDS/files/13025174/1.zip)

**Comment by hong12234** [comment]
![image](https://github.com/eProsima/Fast-DDS/assets/38304396/85052c50-0f15-4e7b-8700-caa41bff0f38)
Hello, I have caught both normal and abnormal network packets from both parties. The data type analysis is the same, but Fastdds cannot receive the data，Look at the picture,This is the log captured by sending and replying to the compound medicine separately.
![image](https://github.com/eProsima/Fast-DDS/assets/38304396/63f33867-76d2-4c1b-b916-2302b5440d7d)

**Comment by hong12234** [comment]
I still don't understand a question, why does fastdds reply with a dimainid of 115 instead of 1.Next, we will upload the network package。
![image](https://github.com/eProsima/Fast-DDS/assets/38304396/abd3fc0b-10b6-47dd-9e46-3a360e764d68)

**Comment by hong12234** [comment]
[log.zip](https://github.com/eProsima/Fast-DDS/files/13038299/log.zip)

**Comment by hong12234** [comment]
![image](https://github.com/eProsima/Fast-DDS/assets/38304396/650e4f09-da14-4b3d-b532-f6295d995b4d)
This is the code we receive data from

**Comment by JesusPoderoso** [comment]
Hi @hong12234 can you confirm if this is happening to you with Fast DDS v2.12.0? This interoperability issue was solved in that release.

**Comment by hong12234** [comment]
![image](https://github.com/eProsima/Fast-DDS/assets/38304396/dd2db381-e456-49ab-8c1b-e68f0c11b26d)
ban'b版本是这个，现在我尝试用2.12测试

**Comment by JesusPoderoso** [comment]
Hi @hong12234 , sorry for the previous message, I was wrong. Despite of Fast CDR v2.0.0 introduces the XCDR2 feature, it is not introduced yet in Fast DDS v2.12.0. The XTypes feature, wich is needed for introducing XCDR2 in Fast DDS, is WIP right now. It will be introduced soon in future releases (see our [roadmap](https://github.com/eProsima/Fast-DDS/blob/master/roadmap.md)), and that will allow  the interoperability with OpenDDS.

**Comment by MiguelCompany** [comment]
@hong12234 As an additional note, setting OpenDDS to use XCDR1 representation would make it interoperate with Fast DDS

**Comment by hong12234** [comment]
> @hong12234作为附加说明，将 OpenDDS 设置为使用 XCDR1 表示形式将使其与 Fast DDS 互操作

Setting OpenDDS to use XCDR1 has already been set, otherwise communication would not be possible. However, now there will be multiple data transmissions from OpenDDS in communication, and most of the received data will be different from the actual sender, or the data sent by OpenDDS cannot be received by FastDDS.
Analyzing the RTPS layer, it can be found that the underlying data fastdds does indeed reply to ACKNACK packets, but cannot take the data. No response.

**Comment by hong12234** [comment]
Refer to this schematic diagram，
![eaf47918f1fbe8801e87d6bd49f125c](https://github.com/eProsima/Fast-DDS/assets/38304396/ae162030-d7cf-4436-906d-4e43cdc2a58c)
This data is two consecutive packets sent by Opendds.
![image](https://github.com/eProsima/Fast-DDS/assets/38304396/2457321b-f28b-4222-9d5d-2ab9e139e079)
It can also be seen from the heartbeat packet that two packets of data have been written，Fastdds also received data.
![image](https://github.com/eProsima/Fast-DDS/assets/38304396/99b8fac6-d0f8-4641-ada5-27b9ded99d1f)
![image](https://github.com/eProsima/Fast-DDS/assets/38304396/acc6ef36-0611-4f66-bcc2-71d6a72e2ec4)
I don't understand why two consecutive pieces of data are sent here.And the first packet of data is still incorrect

**Comment by hong12234** [comment]
Here, Oepndds sends a packet of data, analyzing the underlying interaction between heartbeat packets and ACK packets. At this point, it is reasonable to say that fastdds has already received the data, but in reality, fastdds on_data_available No functions were called。
![image](https://github.com/eProsima/Fast-DDS/assets/38304396/0e22e85a-ced7-49e6-ae5d-ca910afaa66e)
![image](https://github.com/eProsima/Fast-DDS/assets/38304396/20fdcfc8-33bf-40e7-8925-1281dfb37f6f)

**Comment by hong12234** [comment]
[ivi_1019_03.zip](https://github.com/eProsima/Fast-DDS/files/13059664/ivi_1019_03.zip)
This is the packet compression package shown in the screenshot above

**Comment by hong12234** [comment]
![image](https://github.com/eProsima/Fast-DDS/assets/38304396/fa757b27-c4bb-43a2-a0c5-ccc53842b109)
This is our data idl setting,Is it possible to interact with the subject_id The is related, and this data type is defined as @ key。
here are approximately 20 types of subject_id,This is my subscriber's initialization
![image](https://github.com/eProsima/Fast-DDS/assets/38304396/e41f71d3-de21-4076-94ec-b8cfb75acb8d)

**Comment by hong12234** [comment]
![image](https://github.com/eProsima/Fast-DDS/assets/38304396/e5646cd5-4f18-42d6-9ebf-fe9d2d52e6f1)
Also, here is my CDR version

**Comment by JLBuenoLopez** [comment]
@hong12234 

Thanks for your report. [Fast DDS roadmap](https://github.com/eProsima/Fast-DDS/blob/master/roadmap.md) includes support for XTypes v1.3 specification which includes complete XCDR encoding version 2 support. The intention is to ensure interoperability once XCDRv2 is supported through the [OMG DDS interoperability repository](https://github.com/omg-dds/dds-rtps) (the interoperabilty test suite is run using XCDRv2). I am moving this issue to the support forum, as more work is still needed to find if there is a bug and in which side. Sorry for not being of help. You might consider contacting [eProsima's commercial support team](https://github.com/eProsima/Fast-DDS/tree/master#getting-help) for premium support options if this is a need for your project.
